crates/extrinsics/src/contract_storage_tests.rs

```diff
@@ -0,0 +1,242 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::contract_storage::{
+    ContractStorageData,
+    ContractStorageLayout,
+};
+use contract_transcode::ContractMessageTranscoder;
+
+use ink::{
+    metadata::{
+        layout::{
+            Layout::{
+                self,
+                Struct,
+            },
+            LayoutKey,
+            RootLayout,
+        },
+        ConstructorSpec,
+        ContractSpec,
+        InkProject,
+        LangError,
+        MessageSpec,
+        ReturnTypeSpec,
+        TypeSpec,
+    },
+    storage::{
+        traits::{
+            ManualKey,
+            Storable,
+            StorageLayout,
+        },
+        Lazy,
+        Mapping,
+    },
+    ConstructorResult,
+    MessageResult,
+};
+
+use scale::Encode;
+use std::collections::BTreeMap;
+use subxt::backend::legacy::rpc_methods::Bytes;
+
+const BASE_KEY_RAW: [u8; 16] = [0u8; 16];
+const ROOT_KEY: u32 = 0;
+const LAZY_TYPE_ROOT_KEY: u32 = 1;
+
+fn contract_default_spec() -> ContractSpec {
+    ContractSpec::new()
+        .constructors(vec![ConstructorSpec::from_label("new")
+            .selector([94u8, 189u8, 136u8, 214u8])
+            .payable(true)
+            .args(vec![])
+            .returns(ReturnTypeSpec::new(TypeSpec::with_name_str::<
+                ConstructorResult<()>,
+            >(
+                "ink_primitives::ConstructorResult"
+            )))
+            .docs(Vec::new())
+            .done()])
+        .messages(vec![MessageSpec::from_label("inc")
+            .selector([231u8, 208u8, 89u8, 15u8])
+            .mutates(true)
+            .payable(true)
+            .args(vec![])
+            .returns(ReturnTypeSpec::new(TypeSpec::with_name_str::<
+                MessageResult<()>,
+            >(
+                "ink_primitives::MessageResult"
+            )))
+            .default(true)
+            .done()])
+        .events(Vec::new())
+        .lang_error(TypeSpec::with_name_segs::<LangError, _>(
+            ::core::iter::Iterator::map(
+                ::core::iter::IntoIterator::into_iter(["ink", "LangError"]),
+                ::core::convert::AsRef::as_ref,
+            ),
+        ))
+        .done()
+}
+
+fn encode_storage_value<T: Storable>(value: &T) -> Bytes {
+    let mut value_encoded = Vec::new();
+    Storable::encode(value, &mut value_encoded);
+    Bytes::from(value_encoded)
+}
+
+#[test]
+fn storage_decode_simple_type_works() {
+    let root_key_encoded = Encode::encode(&ROOT_KEY);
+    #[derive(scale_info::TypeInfo, StorageLayout, Storable)]
+    struct Data {
+        a: i32,
+    }
+
+    if let Struct(data_layout) = <Data as StorageLayout>::layout(&ROOT_KEY) {
```

Comment:
Could use [`let-else` ](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html) to have the panic at the top and remove the indentation of the test body e.g.

```rust
let Struct(data_layout) = <Data as StorageLayout>::layout(&ROOT_KEY) else {
  panic!("Layout shall be created");
}

// use `data_layout` here
```